packages/core/src/shared/plugins/suggestion/SuggestionPlugin.ts

```diff
@@ -1,146 +1,101 @@
-    notFoundCount: 0,
-    decorationId: undefined,
+export type SuggestionsMenuCallbacks<T extends SuggestionItem> =
+  BaseUiElementCallbacks & {
+    // The function to execute when selecting a suggested item.
+    itemCallback: (item: T) => void;
   };
-}
 
-type SuggestionPluginViewOptions<
-  T extends SuggestionItem,
-  BSchema extends BlockSchema
-> = {
-  editor: BlockNoteEditor<BSchema>;
-  pluginKey: PluginKey;
-  onSelectItem: (props: { item: T; editor: BlockNoteEditor<BSchema> }) => void;
-  suggestionsMenuFactory: SuggestionsMenuFactory<T>;
-};
+export type SuggestionsMenuState<T extends SuggestionItem> =
+  BaseUiElementState & {
+    // The suggested items to display.
+    filteredItems: T[];
+    // The index of the suggested item that's currently hovered by the keyboard.
+    keyboardHoveredItemIndex: number;
+  };
 
-class SuggestionPluginView<
+class SuggestionsMenuView<
   T extends SuggestionItem,
   BSchema extends BlockSchema
 > {
   editor: BlockNoteEditor<BSchema>;
+  ttEditor: Editor;
   pluginKey: PluginKey;
 
-  suggestionsMenu: SuggestionsMenu<T>;
+  private suggestionsMenuState?: SuggestionsMenuState<T>;
+  public updateSuggestionsMenu: () => void;
 
   pluginState: SuggestionPluginState<T>;
   itemCallback: (item: T) => void;
 
-  private lastPosition: DOMRect | undefined;
-
-  constructor({
-    editor,
-    pluginKey,
-    onSelectItem: selectItemCallback = () => {},
-    suggestionsMenuFactory,
-  }: SuggestionPluginViewOptions<T, BSchema>) {
+  constructor(
+    editor: BlockNoteEditor<BSchema>,
+    tiptapEditor: Editor,
+    pluginKey: PluginKey,
+    onSelectItem: (props: {
+      item: T;
+      editor: BlockNoteEditor<BSchema>;
+    }) => void = () => {},
+    updateSuggestionsMenu: (
+      suggestionsMenuState: SuggestionsMenuState<T>
+    ) => void = () => {}
+  ) {
     this.editor = editor;
-    this.pluginKey = pluginKey;
+    this.ttEditor = tiptapEditor;
 
+    this.pluginKey = pluginKey;
     this.pluginState = getDefaultPluginState<T>();
 
+    this.updateSuggestionsMenu = () => {
+      if (!this.suggestionsMenuState) {
+        throw new Error("Attempting to update uninitialized suggestions menu");
+      }
+
+      updateSuggestionsMenu(this.suggestionsMenuState);
+    };
+
     this.itemCallback = (item: T) => {
-      editor._tiptapEditor
+      this.ttEditor
         .chain()
         .focus()
         .deleteRange({
           from:
             this.pluginState.queryStartPos! -
             this.pluginState.triggerCharacter!.length,
-          to: editor._tiptapEditor.state.selection.from,
+          to: this.ttEditor.state.selection.from,
         })
         .run();
 
-      selectItemCallback({
+      onSelectItem({
         item: item,
         editor: editor,
       });
     };
 
-    this.suggestionsMenu = suggestionsMenuFactory(this.getStaticParams());
+    document.addEventListener("scroll", this.handleScroll);
   }
 
+  handleScroll = () => {
```

Comment:
why is this necessary in the new version, but not in the old one?

Reply:
We used to provide a function that returns the `referenceRect` in `staticParams`, but this was pretty much just because of quirks with Tippy that we didn't fully understand at the time (basically so it wouldn't trigger an update). However, it makes more sense from an API design perspective to just provide the `referenceRect` in the `dynamicParams`/`state`, especially since now the components aren't forced to update.